### PR TITLE
OSM

### PR DESCRIPTION
**Pull request by hannes44**
Added OSM support

### CODE REVIEW

File: TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/RoadSystem.cs [hunk 1]
```diff
@@ -9,27 +9,66 @@ public enum DrivingSide
...
         [SerializeField] private GameObject _railPrefab;        
         [SerializeField] private GameObject _intersectionPrefab;
-
+        [SerializeField] private MapGenerator _mapGenerator;
         [SerializeField] private GameObject _roadSystemGraphNodePrefab;
+        [SerializeField] private GameObject _roadSideParkingPrefab;
+        [SerializeField] private GameObject _parkingLotPrefab;
 
         [Header("Road system settings")]
         public DrivingSide DrivingSide = DrivingSide.Right;
```
Comment (belgien) [inline TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/RoadSystem.cs:60]: Add space

File: TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/RoadSystem.cs [hunk 2]
```diff
@@ -163,17 +204,100 @@ private bool PositionsAreInRoadSystem(Vector3[] positions)
...
+
+        public void DeleteAllBusStops()
+        {
+            List<GameObject> busStops = new List<GameObject>();
+
+            foreach(Transform busStopT in BusStopContainer.transform)
+                busStops.Add(busStopT.gameObject);
+            foreach(GameObject busStop in busStops)
+                DestroyImmediate(busStop);
+        }
```
Comment (belgien) [inline TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/RoadSystem.cs:288]: Add space

File: TrafficSimulator/Assets/RoadGenerator/Core/Editor/RoadEditor.cs [hunk 3]
```diff
@@ -197,6 +200,17 @@ public override void OnInspectorGUI()
...
+            if (_isOneWay.boolValue != road.IsOneWay)
+            {
+                changed = true;
+                road.IsOneWay = _isOneWay.boolValue;
+            }
+
+            if (_isGeneratingOSM.boolValue != road.IsGeneratingOSM)
+            {
+                road.IsGeneratingOSM = _isGeneratingOSM.boolValue;
+            }
```
Comment (belgien) [inline TrafficSimulator/Assets/RoadGenerator/Core/Editor/RoadEditor.cs:212]: remove brackets

File: TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/FootWayMeshGenerator.cs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+{
+
+    public class FootWayMeshBuilder
+    {
+        enum VerticeType{
+            BottomLeft = 0,
+            TopLeft = 1,
+            BottomRight = 2,
+            TopRight = 3
+
```
Comment (belgien) [inline TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/FootWayMeshGenerator.cs:17]: remove space

File: TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/FootWayMeshGenerator.cs [hunk 5]
```diff
@@ -0,0 +1,130 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using UnityEngine;
+
+namespace RoadGenerator
+{
+
```
Comment (belgien) [inline TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/FootWayMeshGenerator.cs:9]: remove space

File: TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/FootWayMeshGenerator.cs [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        public List<int> Triangles = new List<int>();
+        public List<Vector2> UVS = new List<Vector2>();
+        public List<Vector3> Normals = new List<Vector3>();
+
+        private float _height;
+
+        public FootWayMeshBuilder(float height)
+        {
+            _height = height;
+        }
```
Comment (belgien) [inline TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/FootWayMeshGenerator.cs:29]: Add space

File: TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/FootWayMeshGenerator.cs [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+            FootWayMeshBuilder builder = new FootWayMeshBuilder(height);
+
+            foreach (var innerOuterPoints in InnerOuterPoints)
+                builder.AddNode(innerOuterPoints);
+
+            Mesh mesh = new Mesh();
+            mesh.vertices = builder.Vertices.ToArray();
+            mesh.triangles = builder.Triangles.ToArray();
+            mesh.uv = builder.UVS.ToArray();
+            mesh.normals = builder.Normals.ToArray();
```
Comment (belgien) [inline TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/FootWayMeshGenerator.cs:98]: Add space

File: TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/FootWayMeshGenerator.cs [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+                Vector3 bottomLeft = point1 - perpendicular * width / 2;
+                Vector3 bottomRight = point1 + perpendicular * width / 2;
+                builder.AddNode((bottomLeft, bottomRight));
+            }
+
+            Mesh mesh = new Mesh();
+            mesh.vertices = builder.Vertices.ToArray();
+            mesh.triangles = builder.Triangles.ToArray();
+            mesh.uv = builder.UVS.ToArray();
+            mesh.normals = builder.Normals.ToArray();
```
Comment (belgien) [inline TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/FootWayMeshGenerator.cs:126]: Add space

File: TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Intersection.cs [hunk 9]
```diff
@@ -963,6 +969,12 @@ private LaneNode GetClosestIndexExitNode(List<LaneNode> exitNodes, int index)
         {
             List<GuideNode> guidePaths = GetGuidePaths(current).Select(x => x.Item3).ToList();
             
+            if (guidePaths.Count == 0)
+            {
+                Debug.Log($"No guide paths found for lane node {current.Position}" + current.Type);
+                return (current.Last, current);
+            }
```
Comment (belgien) [inline TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Intersection.cs:976]: Maybe no debug log?

File: TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Intersection.cs [hunk 10]
```diff
@@ -1110,38 +1122,84 @@ public bool IsThreeWayIntersection()
         /// <summary> Sets the opposite arm and the flow group for the arms </summary>
         public void SetupIntersectionArms()
         {
-           foreach (IntersectionArm intersectionArm in IntersectionArms)
+            Debug.Log("Setting up intersection arms for intersection: " + this);
```
Comment (belgien) [inline TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Intersection.cs:1125]: Debug log ?

File: TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Intersection.cs [hunk 11]
```diff
@@ -1110,38 +1122,84 @@ public bool IsThreeWayIntersection()
...
-                    float angle = Vector3.Angle(intersectionArmDirection, directionBetweenTheJunctionEdges);
-                    if (angle < minAngle)
+                    Vector3 positionAtHalfWay = Vector3.Lerp(intersectionArm.JunctionEdgePosition, otherIntersectionArm.JunctionEdgePosition, 0.5f);
+                    float distance = Vector3.Distance(positionAtHalfWay, IntersectionPosition);
+                    if (distance < minDistance)
                     {
-                        minAngle = angle;
+                        minDistance = distance;
                         minAngleArm = otherIntersectionArm;
+
```
Comment (belgien) [inline TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Intersection.cs:1151]: remove space

File: TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Intersection.cs [hunk 12]
```diff
@@ -1110,38 +1122,84 @@ public bool IsThreeWayIntersection()
...
+            {
+                foreach(IntersectionArm intersectionArm in IntersectionArms)
+                {
+                    if (intersectionArm.OppositeArmID == "")
+                        continue;
+                    
+                    if (GetArm(intersectionArm.OppositeArmID).OppositeArmID != intersectionArm.ID)
+                    {
+                        intersectionArm.OppositeArmID = "";
+                    }
```
Comment (belgien) [inline TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Intersection.cs:1190]: no brackets

File: TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/IntersectionCreator.cs [hunk 13]
```diff
@@ -87,13 +87,58 @@ static IntersectionPointData CalculateIntersectionData(Vector3 intersectionPosit
...
+            {
+                junctionEdgeDatas.AddRange(GetIntersectionAnchorPoints(road, intersectionPosition, intersectionLength));
+            }
+
+
+            IntersectionType type = IntersectionType.FourWayIntersection;
+            if (junctionEdgeDatas.Count == 3)
+            {
+                type = IntersectionType.ThreeWayIntersectionAtStart;
+            }
```
Comment (belgien) [inline TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/IntersectionCreator.cs:114]: No brackets

File: TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/IntersectionCreator.cs [hunk 14]
```diff
@@ -87,13 +87,58 @@ static IntersectionPointData CalculateIntersectionData(Vector3 intersectionPosit
...
             return new IntersectionPointData(intersectionPosition, junctionEdgeDatas, type);
```
Comment (belgien) [inline TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/IntersectionCreator.cs:115]: Add space

File: TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/IntersectionCreator.cs [hunk 15]
```diff
@@ -87,13 +87,58 @@ static IntersectionPointData CalculateIntersectionData(Vector3 intersectionPosit
...
+        {
+            List<JunctionEdgeData> junctionEdgeDatas = new List<JunctionEdgeData>();
+            // TODO fix
+            float intersectionLength = Intersection.CalculateIntersectionLength(roads[0], roads[1]);
+            // Get the anchor points on either side of the intersection for the two roads
+
+            foreach(Road road in roads)
+            {
+                junctionEdgeDatas.AddRange(GetIntersectionAnchorPoints(road, intersectionPosition, intersectionLength));
+            }
```
Comment (belgien) [inline TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/IntersectionCreator.cs:107]: No brackets

File: TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/IntersectionCreator.cs [hunk 16]
```diff
@@ -151,11 +196,12 @@ orderby intersectionData.SegmentIndex
...
 
             DeleteAnchorsInsideIntersectionBounds(intersection);
-
             foreach (Road road in roads)
+            {
                 road.OnChange();
-
-            intersection.UpdateMesh();
+            }
+            return intersection;
```
Comment (belgien) [inline TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/IntersectionCreator.cs:203]: Add space

File: TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Road.cs [hunk 17]
```diff
@@ -243,6 +247,12 @@ public void MapConnectedRoads()
                     if (road == currentRoad)
                         continue;
 
+                    if (road.IsOneWay != currentRoad.IsOneWay)
+                        continue;
+
+                   // if (road.IsOneWay)
+                   //     continue;
```
Comment (belgien) [inline TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Road.cs:254]: Remove

File: TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Road.cs [hunk 18]
```diff
@@ -525,11 +560,16 @@ public void OnChange()
             if(RoadSystem == null)
                 return;
 
-            ConnectRoadIfEndPointsAreClose();
+            if (!IsGeneratingOSM)
+            {
+                ConnectRoadIfEndPointsAreClose();
+            }
```
Comment (belgien) [inline TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Road.cs:566]: No brackets

File: TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Road.cs [hunk 19]
```diff
@@ -539,9 +579,7 @@ protected void UpdateRoad()
             
             foreach(Intersection intersection in Intersections)
                 intersection.UpdateMesh();
-            
-            PlaceTrafficSigns();
-
+            //PlaceTrafficSigns();
```
Comment (belgien) [inline TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Road.cs:582]: Remove comment

File: TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Road.cs [hunk 20]
```diff
@@ -554,10 +592,8 @@ public void UpdateRoadNoGraphUpdate()
             UpdateRoadNodes();
             UpdateLanes();
             UpdateMesh();
-            
-            foreach(Intersection intersection in Intersections)
-                intersection.UpdateMesh();
-            
+            // foreach(Intersection intersection in Intersections)
+            //     intersection.UpdateMesh();
```
Comment (belgien) [inline TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Road.cs:596]: Remove comments

File: TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Road.cs [hunk 21]
```diff
@@ -1063,6 +1099,20 @@ protected GameObject SpawnTrafficSign(TrafficSignData data)
...
+            GameObject busStop = Instantiate(busStopPrefab, roadNode.Position, rotation);
+            busStop.name = busStopName;
+            bool isDrivingRight = RoadSystem.DrivingSide == DrivingSide.Right;
+            Vector3 offsetDirection =  roadNode.Normal * (isDrivingRight ? 1 : -1) * (IsForward ? 1 : -1);
+            float distanceFromRoad = 1f;
+            busStop.transform.position += (distanceFromRoad + LaneCount / 2 * LaneWidth) * offsetDirection;
+            TextMesh text = busStop.GetComponentInChildren<TextMesh>();
+            text.text = busStopName;
+            busStop.transform.parent = RoadSystem.BusStopContainer.transform;
+        }
```
Comment (belgien) [inline TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Road.cs:1115]: Maybe cleanup function, add spacing, and add comments

File: TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Road.cs [hunk 22]
```diff
@@ -1134,6 +1184,14 @@ private void UpdatePOIs()
...
         }
 
+        public void AddFullRoadSideParking(LaneSide laneSide)
+        {
+            GameObject roadsideParkingObject = Instantiate(RoadSystem.RoadSideParkingPrefab, _POIContainer.transform);
+            RoadsideParking roadsideParking = roadsideParkingObject.GetComponent<RoadsideParking>();
+            roadsideParking.LaneSide = laneSide;
+            roadsideParking.Road = this;
+            roadsideParking.ParkingType = RoadSideParkingType.FullRoad;
+        }
```
Comment (belgien) [inline TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Road.cs:1194]: Add comments maybe

File: TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Road.cs [hunk 23]
```diff
@@ -1432,8 +1490,8 @@ void OnDestroy()
                 DestroyImmediate(intersection.gameObject);
             }
 
-            if(!_isBeingDestroyed)
-                RoadSystem.UpdateRoadSystemGraph();
+           // if(!_isBeingDestroyed)
+           //     RoadSystem.UpdateRoadSystemGraph();
```
Comment (belgien) [inline TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Road.cs:1494]: Remove comment

**Review by marc7s — APPROVED** [review]

**Pull request merged**